File: dart/triangle.dart
class Triangle {
  bool isValid(double a, double b, double c) {
    return a > 0 && b > 0 && c > 0 && a + b >= c && a + c >= b && b + c >= a;
  }

  bool equilateral(double a, double b, double c) {
    return isValid(a, b, c) && a == b && b == c;
  }

  bool isosceles(double a, double b, double c) {
    return isValid(a, b, c) && (a == b || b == c || a == c);
  }

  bool scalene(double a, double b, double c) {
    return isValid(a, b, c) && !isosceles(a, b, c);
  }
}

File: lua/triangle.lua
local triangle = {}

function triangle.kind(a, b, c)
    assert(valid(a, b, c), "Input Error")
    if a == b and b == c then
        return "equilateral"
    end

    if a == b or a == c or b == c then
        return "isosceles"
    end
    return "scalene"
end

function valid(a, b, c)
    return a > 0 and b > 0 and c > 0 and a + b >= c and b + c >= a and a + c >= b
end

return triangle
